Refactor cssText() to share more code between grouping at-rules

mdubet opened the pull request:
#### f2c69208138b9fc7b202db0dba43a43696e4a22e
<pre>
Refactor cssText() to share more code between grouping at-rules
<a href="https://bugs.webkit.org/show_bug.cgi?id=251633">https://bugs.webkit.org/show_bug.cgi?id=251633</a>
rdar://104977959

Reviewed by Antti Koivisto.

This is a pure refactoring without any behavioral change.

* Source/WebCore/css/CSSContainerRule.cpp:
(WebCore::CSSContainerRule::cssText const):
* Source/WebCore/css/CSSGroupingRule.cpp:
(WebCore::CSSGroupingRule::appendCSSTextForItems const):
(WebCore::CSSGroupingRule::cssTextForDeclsAndRules const):
* Source/WebCore/css/CSSGroupingRule.h:
* Source/WebCore/css/CSSLayerBlockRule.cpp:
(WebCore::CSSLayerBlockRule::cssText const):
* Source/WebCore/css/CSSMediaRule.cpp:
(WebCore::CSSMediaRule::cssText const):
* Source/WebCore/css/CSSSupportsRule.cpp:
(WebCore::CSSSupportsRule::cssText const):
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::consumeDeclarationListOrStyleBlockHelper):

Canonical link: <a href="https://commits.webkit.org/259815@main">https://commits.webkit.org/259815@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bcf3886ea0094250d4de26d37edc4a9ed7756e0f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106030 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15086 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38862 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115212 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/175315 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/109933 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16510 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6263 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98239 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/114943 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111783 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/12570 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95542 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/40110 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94435 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27185 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/81787 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8339 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28537 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/8830 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/5092 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14452 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48080 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6792 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10374 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/css/CSSGroupingRule.cpp:
```diff
@@ -110,10 +111,51 @@ ExceptionOr<void> CSSGroupingRule::deleteRule(unsigned index)
 }
 
-void CSSGroupingRule::appendCSSTextForItems(StringBuilder& result) const
+String CSSGroupingRule::cssText() const
```
It is confusing to call this just `cssText`  as it is a helper that never matches .cssText output. 

It would be more efficient if this took StringBuilder as a parameter and just wrote into it. Maybe `appendCSSTextForChildRulesAndDeclarations` or similar?

mdubet replied:
Changed

anttijk commented on Source/WebCore/css/CSSGroupingRule.cpp:
```diff
@@ -110,10 +111,51 @@ ExceptionOr<void> CSSGroupingRule::deleteRule(unsigned index)
+    builder.append('\n');
+    builder.append(decls);
+    builder.append(rules);
+    builder.append("\n}");
```
You can give multiple parameters to append and do this in a single line.

mdubet replied:
I think there is no overload of `append()` which takes a mix of StringBuilder and String. Maybe it's worth adding one though ?

mdubet replied:
There is an overload with StringBuilder&& though which is more efficient

mdubet replied:
Actually no it's a FIXME to support StringBuilder : https://github.com/WebKit/WebKit/blob/main/Source/WTF/wtf/text/StringBuilder.h#L70

anttijk commented on Source/WebCore/css/CSSGroupingRule.cpp:
```diff
@@ -110,10 +111,51 @@ ExceptionOr<void> CSSGroupingRule::deleteRule(unsigned index)
+
+    builder.append(" {");
+
+    auto [decls, rules] = cssTextForItems();
```
`declarations`

mdubet replied:
`decls` and `rules` are spec vocabulary (cf https://www.w3.org/TR/cssom/#css-rules)

anttijk commented on Source/WebCore/css/CSSGroupingRule.h:
```diff
@@ -46,7 +46,14 @@ class CSSGroupingRule : public CSSRule {
+        StringBuilder decls;
+        StringBuilder rules;
+    };
+    CSSText cssTextForItems() const;
```
Returning StringBuilders is odd. They are only used to extract strings so they could just be Strings.

It would be more efficient to provide StringBuilders as arguments that cssTextForItems would write to.

"Items" is vague.

mdubet replied:
`items` is also CSSOM spec vocabulary.

I will change to pass StringBuilder as arguments.

mdubet committed:
Refactor cssText() to share more code between grouping at-rules
https://bugs.webkit.org/show_bug.cgi?id=251633
rdar://104977959

Reviewed by Antti Koivisto.

This is a pure refactoring without any behavioral change.

* Source/WebCore/css/CSSContainerRule.cpp:
(WebCore::CSSContainerRule::cssText const):
* Source/WebCore/css/CSSGroupingRule.cpp:
(WebCore::CSSGroupingRule::appendCSSTextForItems const):
(WebCore::CSSGroupingRule::cssTextForDeclsAndRules const):
* Source/WebCore/css/CSSGroupingRule.h:
* Source/WebCore/css/CSSLayerBlockRule.cpp:
(WebCore::CSSLayerBlockRule::cssText const):
* Source/WebCore/css/CSSMediaRule.cpp:
(WebCore::CSSMediaRule::cssText const):
* Source/WebCore/css/CSSSupportsRule.cpp:
(WebCore::CSSSupportsRule::cssText const):
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::consumeDeclarationListOrStyleBlockHelper):

Canonical link: https://commits.webkit.org/259815@main
